Add identity events for Git commits

jjmerchante opened the pull request:
This PR introduces functionality to extract and eventize identities from Git commit data. It handles authors, committers, and signers, creating events for each identity found.

jjmerchante committed:
[git] Add identity events for Git commits

This commit introduces functionality to extract and
eventize identities from Git commit data. It handles
authors, committers, and signers, creating events for
each identity found.

Signed-off-by: Jose Javier Merchante <jjmerchante@bitergia.com>

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -32,6 +33,28 @@
+    "reviewed-by",
+    "signed-off-by",
+    "suggested-by",
+    "tested-by",
```
I think Perceval should be responsible for finding the trailers, and chronicler for figuring out the identities inside them.

sduenas replied:
Let's use as supported trailers the ones listed by the Git documentation:

- Reported-by: is used to credit someone who found the bug that the patch attempts to fix.
- Acked-by: says that the person who is more familiar with the area the patch attempts to modify liked the patch.
- Reviewed-by:, unlike the other trailers, can only be offered by the reviewers themselves when they are completely satisfied with the patch after a detailed analysis.
- Tested-by: is used to indicate that the person applied the patch and found it to have the desired effect.
- Co-authored-by: is used to indicate that people exchanged drafts of a patch before submitting it.
- Helped-by: is used to credit someone who suggested ideas for changes without providing the precise changes in patch form.
- Mentored-by: is used to credit someone with helping develop a patch as part of a mentorship program (e.g., GSoC or Outreachy).
- Suggested-by: is used to credit someone with suggesting the idea for a patch.

Plus these ones that are common on the Linux kernel commits:

- Reported-by
- Mentored-by

Reference: https://lore.kernel.org/git/60ad75ac7ffca_2ae08208b@natae.notmuch/

sduenas commented on tests/test_git.py:
```diff
@@ -25,7 +25,8 @@
                                         GIT_EVENT_ACTION_REPLACED,
-                                        GIT_EVENT_ACTION_COPIED)
+                                        GIT_EVENT_ACTION_COPIED,
+                                        IDENTITY_EVENT_TYPE)
```
Instead of calling it `IDENTITY_EVENT_TYPE` what I would do is to create events for all the trailers, something like `GIT_EVENT_COMMIT_APPROVED_BY`, `GIT_EVENT_COMMIT_MERGED_BY`, and so on.

In the future, we might consider to make these events generic for other types of datasources.

sduenas replied:
New proposal with a set of new events.

sduenas commented on chronicler/eventizer.py:
```diff
@@ -157,3 +157,83 @@ def check_value(v):
+    sha1 = hashlib.sha1(s)
+    uuid_ = sha1.hexdigest()
+
+    return uuid_
```
I think we're start having this in many places to avoid a dependency with sortinghat. Why don't we move it to `toolkit` so it's the standard way to create uuids for identities in grimoirelab?

sduenas replied:
It has been added to toolkit.

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -32,6 +35,36 @@
 GIT_EVENT_ACTION_TYPE_CHANGED = "org.grimoirelab.events.git.file.typechanged"
 
+GIT_EVENT_COMMIT_AUTHOR = "org.grimoirelab.events.git.author"
+GIT_EVENT_COMMIT_COMMITTER = "org.grimoirelab.events.git.committer"
```
To keep them aligned with the other events I suggest this change:

```suggestion
GIT_EVENT_COMMIT_AUTHORED_BY = "org.grimoirelab.events.git.commit.authored_by"
GIT_EVENT_COMMIT_COMMITTED_BY = "org.grimoirelab.events.git.commit.committed_by"
```

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -32,6 +35,36 @@
+}
+
+AUTHOR_P2P_REGEX = re.compile(
+    r"(?P<first_authors>.* .*) ([aA][nN][dD]|&|\+) (?P<last_author>.* .*) (?P<email>.*)"
```
Can we find a better regex? This one looks not helpful because we aren't capturing the email with the expression.

Also, I think we need some good examples about how people use these fields. I remember we added peer programming support on ELK long ago. Maybe you can find where we introduced it and look for the examples that we used for that.

jjmerchante replied:
I used the regular expression from ELK: https://github.com/chaoss/grimoirelab-elk/blob/main/grimoire_elk/enriched/git.py#L81.

It looks like there is a single email in the format `name one and name two <foo@bar>`. One example is this commit: https://github.com/mezuro/kalibro/commit/a9f25bea3cee46f9ecae3fd662b79fb4734f1b68.

In ELK the email is ignored. I don’t know who I should assign the email to. To both of them?

The [second regular expression from ELK](https://github.com/chaoss/grimoirelab-elk/blob/main/grimoire_elk/enriched/git.py#L82) is incorrect; it obtains `Co-author` from `Author` and `Commit` fields, but `Co-author` is in the Commit message, not in the `Author` field.

sduenas replied:
The original project we used to incorporate pair programming was cloudfoundy. Maybe you can have some more examples there to add to the tests.

I suggest that the email address should be part of the regular expression, maybe as an optional field. If you do this, you won't have to look for the email address later.

jjmerchante replied:
I can see more examples in this archived repository: https://github.com/cloudfoundry/cflinuxfs2. They follow the same pattern: `Name Surname, Name Surname and Name Surname <oneemail@domain.com>`. The email sometimes comes from the first person and other times from the last.

I have improved the regular expression. The email is detected using `<>` and it is added as a new identity (without a name and username).

I included a test with that pattern.

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -143,3 +181,142 @@ def _process_action(self, source, time, event_uuid, action, file_data):
+
+        events = []
+
+        author = raw_item["data"].get("Author", None)
```
Is it possible that `Author` or `Commit` are empty? That we don't have that field comming from perceval?

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -143,3 +181,142 @@ def _process_action(self, source, time, event_uuid, action, file_data):
+                                                  source_uri=parent_event['source'],
+                                                  time=parent_event['time'],
+                                                  event_type=GIT_EVENT_COMMIT_AUTHOR,
+                                                  event_uuid=parent_event["id"])
```
Can we follow the same pattern/style we were doing with actions? Something like `_process_identities`? I would like to have the same style of processing data in for the same eventizer than two different things.

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -143,3 +181,142 @@ def _process_action(self, source, time, event_uuid, action, file_data):
+            yield self._get_sh_identity(user_str)
+
+    @staticmethod
+    def _get_sh_identity(git_user: str) -> dict:
```
We shouldn't mention sortinghat here. I would call this `_parse_identity`

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -143,3 +181,142 @@ def _process_action(self, source, time, event_uuid, action, file_data):
+            email = git_user.split("<")[1][:-1]
+        identity["username"] = None
+        identity["email"] = email
+        identity["name"] = name
```
It would probably make sense to create a namedtuple for this data structure. I think this structure would be shared by all other eventizers.

jjmerchante replied:
I included the namedtuple in the eventizer because I didn’t see another place for it. I’m not sure if it would be better to create a utils module for that.

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -143,3 +181,142 @@ def _process_action(self, source, time, event_uuid, action, file_data):
+
+        if not any([name, email, username]):
+            logger.warning(f"Identity for event '{event_uuid}' has no name, email or username. Skipping.")
+            return None
```
Probably is better to raise a defined exception and that the upper method captures and processes it. Then we avoid to have to handle `None` cases, that who know what that means.

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -32,6 +35,36 @@
+GIT_EVENT_COMMIT_REVIEWED_BY = "org.grimoirelab.events.git.reviewed_by"
+GIT_EVENT_COMMIT_SIGNED_OFF_BY = "org.grimoirelab.events.git.signed_off_by"
+GIT_EVENT_COMMIT_SUGGESTED_BY = "org.grimoirelab.events.git.suggested_by"
+GIT_EVENT_COMMIT_TESTED_BY = "org.grimoirelab.events.git.tested_by"
```
I think we should add the subdomain `.commit.`  in between `.git.` and the event name to make clear what it is. Same as we have for `file`.

jjmerchante committed:
[ci] Continue workflow when Coveralls fail

This commit updates the tests workflow to continue
when the coveralls action fails as proposed by one
of their incidents:

https://status.coveralls.io/incidents/v5mcbrsbhgt4

Signed-off-by: Jose Javier Merchante <jjmerchante@bitergia.com>

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -143,3 +181,114 @@ def _process_action(self, source, time, event_uuid, action, file_data):
+                                                   time=parent_event['time'],
+                                                   event_uuid=parent_event['id'],
+                                                   event_type=GIT_EVENT_COMMIT_AUTHORED_BY,
+                                                   raw_identities=authors)
```
The parameters of `_process_identities` aren't optional. Remove the `=` when passing them.

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -143,3 +181,114 @@ def _process_action(self, source, time, event_uuid, action, file_data):
+        :returns: generator of CloudEvent with the identity information
+        """
+        for raw_identity in raw_identities:
+
```
```suggestion
```

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -16,12 +16,15 @@
 
-from ...eventizer import Eventizer, uuid
+from cloudevents.http import CloudEvent
+from grimoirelab_toolkit.identities import generate_uuid
```
In the module `eventizer` we have a function called `uuid`. Not sure if there are exactly the same. If they are, I think we should unify them.

jjmerchante replied:
It is not the same. For the `uuid()` function, the arguments must not be None. For `generate_uuid()`, some arguments may be None, but not all. If an argument is None, its None value is used to build the UUID.

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -32,11 +35,41 @@
 GIT_EVENT_ACTION_COPIED = "org.grimoirelab.events.git.file.copied"
 GIT_EVENT_ACTION_TYPE_CHANGED = "org.grimoirelab.events.git.file.typechanged"
 
+GIT_EVENT_COMMIT_AUTHORED_BY = "org.grimoirelab.events.git.commit.authored_by"
+GIT_EVENT_COMMIT_COMMITTED_BY = "org.grimoirelab.events.git.commit.committed_by"
+GIT_EVENT_COMMIT_ACKED_BY = "org.grimoirelab.events.git.commit.acked_by"
+GIT_EVENT_COMMIT_CO_AUTHORED_BY = "org.grimoirelab.events.git.commit.co_authored_by"
+GIT_EVENT_COMMIT_HELPED_BY = "org.grimoirelab.events.git.commit.helped_by"
+GIT_EVENT_COMMIT_MENTORED_BY = "org.grimoirelab.events.git.commit.mentored_by"
+GIT_EVENT_COMMIT_REPORTED_BY = "org.grimoirelab.events.git.commit.reported_by"
+GIT_EVENT_COMMIT_REVIEWED_BY = "org.grimoirelab.events.git.commit.reviewed_by"
+GIT_EVENT_COMMIT_SIGNED_OFF_BY = "org.grimoirelab.events.git.commit.signed_off_by"
+GIT_EVENT_COMMIT_SUGGESTED_BY = "org.grimoirelab.events.git.commit.suggested_by"
+GIT_EVENT_COMMIT_TESTED_BY = "org.grimoirelab.events.git.commit.tested_by"
```
Please document these events. You can add the description and parameters in the documentation folder.
Use as reference https://github.com/cdevents/spec/blob/v0.4.1/core.md

sduenas commented on docs/events.md:
```diff
@@ -110,3 +110,61 @@ File objects found in the `files` field of [commit](#commit) and [merge commit](
+| email | `String` | Email of the contributor                          |
+| uuid  | `String` | Unique identifier of the contributor              |
+| role  | `String` | Role of the contributor.                          |
+| source | `String` | Source of the identity. Always `git` for commits. |
```
Can you align the text and format of the table so it can be read when not rendered with a markdown reader?

sduenas commented on docs/events.md:
```diff
@@ -110,3 +110,61 @@ File objects found in the `files` field of [commit](#commit) and [merge commit](
+
+##### Committer
+
+- Event type: `org.grimoirelab.events.git.commit.committed_by`
+
+##### Acked-by
+
+- Event type: `org.grimoirelab.events.git.commit.acked_by`
+
+##### Co-authored-by
+
+- Event type: `org.grimoirelab.events.git.commit.co_authored_by`
+
+##### Helped-by
+
+- Event type: `org.grimoirelab.events.git.commit.helped_by`
+
+##### Mentored-by
+
+- Event type: `org.grimoirelab.events.git.commit.mentored_by`
+
+##### Reported-by
+
+- Event type: `org.grimoirelab.events.git.commit.reported_by`
+
+##### Reviewed-by
+
+- Event type: `org.grimoirelab.events.git.commit.reviewed_by`
+
+##### Signed-off-by
+
+- Event type: `org.grimoirelab.events.git.commit.signed_off_by`
+
+##### Suggested-by
+
+- Event type: `org.grimoirelab.events.git.commit.suggested_by`
+
+##### Tested-by
+
+- Event type: `org.grimoirelab.events.git.commit.tested_by`
```
I think it would be nice if we add a description of these events. I think  you can add more or less the same text that is available in the git book.

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -143,3 +181,116 @@ def _process_action(self, source, time, event_uuid, action, file_data):
+            authors = m.group("first_authors").split(",")
+            authors = [author.strip() for author in authors]
+            authors += [m.group("last_author")]
+            authors += [f"<{m.group('email')}>"]
```
I think the email address should be shared by all the authors. With the patterns we have seen, it's impossible to know whether it belongs to the first author, second, etc. Sometimes, they even share an email address. I think the safest thing is to assign it to all the identities at the same time.

If these events are handled later, the component that process them show be aware that there's more than one author are sharing an account.

jjmerchante replied:
The email belongs to one of them, as I can see in many commits. This causes the email to be assigned to someone it doesn’t actually belong to. We then have two identities sharing the same email but with different names. If we merge them by email, we’ll end up with a single individual in the commit history.

sduenas replied:
That's normally the case and that could happen, the problem is we don't know whose email is it from the list of identities listed on it. However, I understand what you need let's keep the email as a separate individual.

sduenas commented on chronicler/events/core/git.py:
```diff
@@ -32,11 +35,41 @@
+    "Tested-by": GIT_EVENT_COMMIT_TESTED_BY,
+}
+
+AUTHOR_P2P_REGEX = re.compile(
+    r"(?P<first_authors>.+?)\s+(?:[aA][nN][dD]|&|\+)\s+(?P<last_author>.+?)\s+<(?P<email>[^>]+)>"
+)
```
```suggestion
# Pair programming regex. Some matching examples are:
#   - John Smith, John Doe and Jane Rae <pairprogramming@example.com>
#   - John Smith, John Doe & Jane Rae <pairprogramming@example>
#   - John Smith and John Doe <pairpogramming@example>
GIT_AUTHORS_REGEX = re.compile(
    r"(?P<first_authors>.+?)\s+(?:[aA][nN][dD]|&|\+)\s+(?P<last_author>.+?)\s+<(?P<email>[^>]+)>"
)
```

sduenas reviewed (changes requested):
The PR looks good. I'm suggesting minor changes, mainly in the documentation.

sduenas reviewed (approved):
LGTM
